add autoreload to `trunk serve`

kcking opened the pull request:
Thanks for building trunk! It's such an awesome tool. 

I took a pass at implementing a simple autoreload that is triggered by an injected websocket script. It is exposed by a `--autoreload` flag that defaults to false. It looks like my rustfmt did some collateral formatting in a couple of files, I can manually revert it if need be.

Let me know what you think!

<!--
Thank you for taking the time to open a pull request!
Please review the checklist below and perform each of
the applicable tasks. ❤️!
-->
**Checklist**
- [x] Updated CHANGELOG.md describing pertinent changes.
- [x] Updated README.md with pertinent info (may not always apply).
- [x] Squash down commits to one or two logical commits which clearly describe the work you've done. If you don't, then Dodd will `:)`.

thedodd commented on src/autoreload.js:
```diff
@@ -0,0 +1,9 @@
+(function () {
+    var ws = new WebSocket('ws://' + window.location.host);
```
We should make this connect to the dev server at a location which is not likely to conflict with users' proxy definitions. Something like `'ws://' + window.location.host + '/_trunk/ws'`.

When configured to inject the autoreloader, we can listen for WS connections at that location. Thoughts?

thedodd commented on src/config/models.rs:
```diff
@@ -61,6 +61,10 @@ pub struct ConfigOptsServe {
+    /// Whether to auto-reload the web page when a build completes.
+    #[structopt(long)]
+    #[serde(default)]
+    pub autoreload: bool,
```
Let's definitely make this default to `true`. Folks are going to want it enabled far more often than not.

thedodd commented on src/watch.rs:
```diff
@@ -25,11 +25,13 @@ pub struct WatchSystem {
     /// The watch system used for watching the filesystem.
     _watcher: (JoinHandle<()>, RecommendedWatcher),
+    /// Sender to signal when a build has finished and the page can be refreshed
+    build_done_tx: Option<Sender<()>>,
```
For now, this is good. Long-term we will probably want to update the injected code to include a light UI which is able to describe the error directly in a UI overlay. At that point in time, we can update this signal to carry additional info on any build related errors.

thedodd commented on src/serve.rs:
```diff
@@ -4,8 +4,13 @@ use std::sync::Arc;
 use anyhow::Result;
 use async_std::fs;
 use async_std::task::{spawn, spawn_local, JoinHandle};
+use futures::channel::mpsc::{channel, Receiver};
+use futures::lock::Mutex;
+use futures::{StreamExt, TryFutureExt};
```
I wish we would have already cut over to tokio. Tokio's broadcast channels would be much better for this use case. A single broadcast sender from the WatchSystem. Any number of WebSocket connections could then be open and be subscribed to the signal.

thedodd commented on src/serve.rs:
```diff
@@ -42,19 +51,51 @@ impl ServeSystem {
         Ok(())
     }
 
+    async fn build_notify_task(mut build_done_rx: Receiver<()>, websockets: WebSockets) {
+        while let Some(_) = build_done_rx.next().await {
+            let mut ws_guard = websockets.lock().await;
+            let mut to_remove = vec![];
+            for (idx, ws) in ws_guard.iter_mut().enumerate() {
+                if ws.send_string(r#"{"reload": true}"#.to_string()).await.is_err() {
+                    to_remove.push(idx);
+                }
+            }
+            for idx in to_remove.into_iter().rev() {
+                ws_guard.remove(idx);
+            }
+        }
+    }
```
Once we cut over to tokio, we can simply clone the broadcaster held in the server state, spawn a task which listens for the build completion signals & also listen for the WebSocket to close. That would simplify the setup here quite a lot, and would remove the need for any synchronization (mutex).

For now, solid work `:)`!

thedodd commented on src/serve.rs:
```diff
@@ -96,10 +140,45 @@ impl ServeSystem {
+
+fn inject_reload_script(html: &Vec<u8>) -> Option<Vec<u8>> {
+    let mut script = std::io::Cursor::new(format!("<script>{}</script>", RELOAD_SCRIPT).as_bytes().to_vec());
+    let parsed_script = kuchiki::parse_html().from_utf8().read_from(&mut script).unwrap();
```
After cutting over to nipper, this unwrap may no longer be applicable, but if it is, let's remove it in favor of just logging an error in the console & then returning `None`.

thedodd reviewed (changes requested):
@kcking thanks for the PR! Stoked to see this feature being implemented. A few initial comments, let me know what you think.

kcking commented:
Thanks for the review @thedodd! All your comments sgtm, I'll address them shortly.

kcking commented:
Ok just addressed those comments, let me know if there is anything else!

I changed the flag name from `autoreload` to `no-autoreload` since it is now enabled by default. This seemed simpler than using `--autoreload=false` to disable it, but let me know if you think there's a better way.

thedodd commented:
@kcking sorry to make your PR more difficult! The Tokio cut over is complete, so there are a few conflicts and such. I do think the code will be a bit more simple now though.

On a high level, given Tokio's significantly better messaging support, we should probably use a `tokio::sync::watch` channel, which new WebSocket connections will simply subscribe to and then listen for changes on the channel. For this initial autoreload support, I don't think we need to propagate compiler error messages or anything like that. We can just emit a new value over the channel upon a successful build.

In the future, we can actually propagate a structure message over the websockets so that we can create a diagnostics overlay or drop the info in the console and such.

kcking commented:
Sounds good! I should have time tonight to fix this pr up

Sent via Superhuman iOS ***@***.***>


On Fri, Aug 6 2021 at 8:07 AM, Anthony Dodd ***@***.***>
wrote:

> @kcking <https://github.com/kcking> sorry to make your PR more difficult!
> The Tokio cut over is complete, so there are a few conflicts and such. I do
> think the code will be a bit more simple now though.
>
> On a high level, given Tokio's significantly better messaging support, we
> should probably use a tokio::sync::watch channel, which new WebSocket
> connections will simply subscribe to and then listen for changes on the
> channel. For this initial autoreload support, I don't think we need to
> propagate compiler error messages or anything like that. We can just emit a
> new value over the channel upon a successful build.
>
> In the future, we can actually propagate a structure message over the
> websockets so that we can create a diagnostics overlay or drop the info in
> the console and such.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thedodd/trunk/pull/191#issuecomment-894325026>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKJIU7KMK57CPBDHNZKVT3T3P3CRANCNFSM47V3B5HQ>
> .
>

kcking commented:
@thedodd alrighty just updated this PR to support axum/tokio. Let me know if there's any other changes on your mind :)

thedodd reviewed (approved):
Awesome! I'm going to merge this in a few hours, I need to go AFK for a bit though. Really excellent work @kcking! Thanks!

Notes for myself:
- Add docs about this feature to the site.
- Add docs for this config var to the default Trunk.toml.
- Add to config cascade.
- Add a todo/note in the code about future emission of error info over the socket.
- Probably move code injector the HTML pipeline.

kcking commented on Cargo.toml:
```diff
@@ -44,7 +44,7 @@ structopt-derive = "0.4"
 tar = "0.4"
 tide = { version = "0.16.0", default-features = false, features = ["h1-server", "sessions", "unstable"] }
-tide-websockets = "0.3.0"
+tide-websockets = "0.4.0"
```
I dont see any `Cargo.toml` diffs in the latest changes: https://github.com/thedodd/trunk/pull/191/files, maybe that was an old commit?

kcking commented:
> 
> Awesome! I'm going to merge this in a few hours, I need to go AFK for a bit though. Really excellent work @kcking! Thanks!
> 

Woo! Thank you for the review and for the awesome tool :)

thedodd commented:
🎉

thedodd commented:
BTW, @kcking you should totally join our discord server so that I can give you a shoutout for this work `:)`!
